src/H5RT.c

```diff
@@ -0,0 +1,700 @@
+#endif
+
+/* Check if two hyper-rectangles specified by (min1, max1) and (min2, max2) intersect */
+/*-------------------------------------------------------------------------
+ * Function:    H5RT_leaf_init
+ *
+ * Purpose:     Initialize an R-tree leaf with dynamic coordinate allocation
+ *              based on the specified rank. The leaf structure itself must
+ *              already be allocated by the caller.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5RT_leaf_init(H5RT_leaf_t *leaf, int rank, uintptr_t record)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    if (!leaf)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid leaf pointer");
+
+    if (rank < 1 || rank > H5S_MAX_RANK)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid rank");
+
+    /* Zero out the leaf structure */
+    memset(leaf, 0, sizeof(H5RT_leaf_t));
+
+    /* Allocate coordinate arrays as single block: 3 * rank * sizeof(hsize_t) */
+    if (NULL == (leaf->_coords = (hsize_t *)calloc(3 * (size_t)rank, sizeof(hsize_t))))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "failed to allocate leaf coordinates");
+
+    /* Set up pointers to sections of the coordinate block */
+    leaf->min = leaf->_coords;
+    leaf->max = leaf->_coords + rank;
+    leaf->mid = leaf->_coords + (2 * rank);
+
+    leaf->rank   = rank;
+    leaf->record = record;
+
+done:
+    if (ret_value < 0 && leaf && leaf->_coords) {
+        free(leaf->_coords);
+        leaf->_coords = NULL;
+        leaf->min     = NULL;
+        leaf->max     = NULL;
+        leaf->mid     = NULL;
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5RT_leaf_init() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5RT_leaf_cleanup
+ *
+ * Purpose:     Clean up a leaf's coordinate arrays. The leaf structure
+ *              itself is not freed - that's the caller's responsibility.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5RT_leaf_cleanup(H5RT_leaf_t *leaf)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    if (!leaf)
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "invalid leaf");
+
+    /* Free coordinate arrays */
+    if (leaf->_coords) {
+        free(leaf->_coords);
+        leaf->_coords = NULL;
+        leaf->min     = NULL;
+        leaf->max     = NULL;
+        leaf->mid     = NULL;
+    }
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5RT_leaf_cleanup() */
+
+bool
+H5RT__leaves_intersect(int rank, hsize_t min1[], hsize_t max1[], hsize_t min2[], hsize_t max2[])
+{
+    FUNC_ENTER_PACKAGE_NOERR
+
+    for (int i = 0; i < rank; i++)
+        if (min1[i] > max2[i] || min2[i] > max1[i])
+            return false; /* No overlap in i-th dimension */
```

Comment:
I don't think it's appropriate to bypass FUNC_ENTER/LEAVE with a return here. I would hope modern compilers would handle HGOTO_DONE(false) or just setting a return value just as efficiently.